src/app/components/category-products/category-products.component.ts

```diff
@@ -1,10 +1,172 @@
   encapsulation: ViewEncapsulation.None,
-  changeDetection: ChangeDetectionStrategy.OnPush
+  changeDetection: ChangeDetectionStrategy.OnPush,
 })
 export class CategoryProductsComponent {
+  readonly sortOption: FormControl = new FormControl('fvDesc');
+  readonly filterByPrice: FormGroup = new FormGroup({
+    priceFrom: new FormControl(),
+    priceTo: new FormControl(),
+  });
+
+  readonly sortOptions$: Observable<SortProductsOptionsQueryModel[]> = of([
+    { name: 'Featured', value: 'fvDesc' },
+    { name: 'Price: Low to High', value: 'priceAsc' },
+    { name: 'Price: High to Low', value: 'priceDesc' },
+    { name: 'Avg. Rating', value: 'rvDesc' },
+  ]);
+  readonly pageSizes$: Observable<number[]> = of([5, 10, 15]);
+
+  readonly allCategories$: Observable<CategoryModel[]> =
+    this._categoriesService.getAllCategories();
+
+  readonly categoryDetails$: Observable<CategoryModel> =
+    this._activatedRoute.params.pipe(
+      switchMap((params) =>
+        this._categoriesService.getOneCategory(params['categoryId'])
+      )
+    );
+
+  readonly allProductsInCategory$: Observable<ProductModel[]> = combineLatest([
+    this._activatedRoute.params,
+    this._productsService.getAllProducts(),
+  ]).pipe(
+    map(([params, products]: [Params, ProductModel[]]) =>
+      products.filter((product) => product.categoryId === params['categoryId'])
+    ),
+    shareReplay(1)
+  );
+
+  readonly pageParams$: Observable<PageParams> =
+    this._activatedRoute.queryParams.pipe(
+      map((params) => ({
+        pageNumber: params['pageNumber']
+          ? Math.max(1, +params['pageNumber'])
+          : 1,
+        pageSize: params['pageSize'] ? Math.max(5, +params['pageSize']) : 5,
+      })),
+      shareReplay(1)
+    );
+
+  readonly pageNumbers$: Observable<number[]> = combineLatest([
+    this.allProductsInCategory$,
+    this.pageParams$,
+  ]).pipe(
+    map(([products, params]) => {
+      return Array.from(
+        Array(Math.ceil(products.length / params.pageSize)).keys()
+      ).map((n) => n + 1);
+    })
+  );
+
+  readonly products$: Observable<ProductModel[]> = combineLatest([
+    this.allProductsInCategory$,
+    this.sortOption.valueChanges.pipe(startWith('fvDesc')),
+    this.filterByPrice.valueChanges.pipe(
+      startWith({ priceFrom: 0, priceTo: 9999 }),
+      shareReplay(1)
+    ),
+    this.pageParams$,
+  ]).pipe(
+    map(([products, sortOption, filterByPrice, params]) =>
+      products
+        .slice(
+          (params.pageNumber - 1) * params.pageSize,
+          params.pageNumber * params.pageSize
+        )
+        .sort((a, b) => {
+          if (sortOption === 'fvDesc') {
+            return a.featureValue < b.featureValue ? 1 : -1;
+          }
+
+          if (sortOption === 'priceAsc') {
+            return a.price > b.price ? 1 : -1;
+          }
+
+          if (sortOption === 'priceDesc') {
+            return a.price < b.price ? 1 : -1;
+          }
+
+          if (sortOption === 'rvDesc') {
```

Comment:
If you can, try to do it without using so many ifs.

Reply:
I'm keep trying, so far with no success :)